Sso only

dan-palmer opened the pull request:
- [x]  remove devise log in screens
- [x]  remove old fields on user table
~~- [ ]  tidy up sign in and login spec helper methods~~

Commit:
Introduce a session controller

End goal here is to remove Devise form the app. User management will be
handled by the SSO. We may still want to have persisted users to hook
up to the workflow models and not carry the various user data in the
session.

Utilise the existing omniauth hash helper on the User model to set the
user reference in the session.

The new session template models the current users/sign_in template in
terms of having the same link to the SSO. This landing page might not
be necessary however it should help get the existing specs to pass easily.

Commit:
Configure Omniauth for MoJ SSO

Provides an on failure handler which redirects to to the new session
path for now.

We also need to set the callback path and hook it up to the sessions
controller.

Commit:
Override App Controller methods provided by Devise

Commit:
Remove Devise

Commit:
Update sign out path

Commit:
Update specs to use sessions controller

Commit:
Remove respond_to jpg from images controller

respond_to has been removed from Rails this was only working because
Devise had included the responders gem.

Commit:
Remove devise related specs

Commit:
Remove unused callbacks controller

Commit:
Remove devise specific views

Commit:
Remove devise locale files

Commit:
Remove dev:prime task

We don’t need to create local users for development now. We can hook to
local sso or we could do something like mock omniauth in development.

Commit:
Remove devise related fields from users table

Commit:
Add a user sign out scenario

This spec will change soon as log out from the SSO is added.

Utilise the OAuthHelper method for configuring the OmniAuth mock.

lostie commented on app/controllers/detainees/images_controller.rb:
```diff
@@ -1,6 +1,5 @@
 module Detainees
   class ImagesController < ApplicationController
-    respond_to :jpg
```
👺 : What's the reason for this change (specially in this PR)  ?

dan-palmer replied:
Should be in the commit message for that change. The respond_to method was only available because of Devise adding in the responders gem.

lostie reviewed (commented):
For development purposes, these changes imply that we will need a local version of the MOJ SSO running.

It's likely that the instructions to setup the application need to be updated to:
- Have MOJ SSO app as an **explicit requirement** to run the app
- Update instructions to setup MOJ SSO app so that, ideally, the user only needs to run a _single_ command that sets the application up and running (if it's a dockerized version or not, it's up for discussion)
- Instructions about the default user that can be used to login into the app

Commit:
Add logged in user to the top of the page

If the user is logged in it shows its name in the top right of the
layout next to the sign out button.

Commit:
Provide a way to mock SSO in development mode

dan-palmer commented on app/models/user.rb:
```diff
@@ -22,4 +22,8 @@ def first_or_create_by_uid_and_provider(auth)
   end
+
+  def full_name
+    [first_name, last_name].select(&:present?).join(' ')
```
A faster way to do this(not that that matters) but with a lil less code is:
```ruby
def full_name
  "#{first_name} #{last_name}".strip
end
```

dan-palmer replied:
bump

Commit:
Sign out from SSO app when signing out from MPS app

Whenever signing out from the MPS app ensure that the sign out is also
done in the Single Sign On app.

Uses the same concept used in the PVB app defining an SSO identity
object that handles the creation/deletion of the user/session data.

Commit:
Ensure working app when using SSO mock mode

When in development mode the app should still be usable. More
specifically, when logging out, the app should redirect to the 'Login as
SSO' screen (instead of trying, unsucessfully, to redirect to the SSO
app login screen).

In test mode the redirect to SSO app doesn't work by default as well.
These changes introduce a dumb rack server that accepts the requests
that would normally end up in the SSO app.

lostie commented:
@dan-palmer I've added an isolated [commit](https://github.com/ministryofjustice/moving-people-safely/pull/248/commits/2e797f2736db3c1e6707cbb4dcb776e2abcea36a) that addresses the issue of logging out in _mock mode_ (using the logout URL contained in the SSO data).

Basically, the approach I took was to redirect the user to the _new_session_path_ in development, and to create a (very basic) mocked SSO app for testing purposes.

Please 👓  ;)

Commit:
Add ENV option on assets precompilation to skip given initializers

dan-palmer commented on app/models/sso/identity.rb:
```diff
@@ -0,0 +1,62 @@
+          permissions: hash.fetch('permissions')
+        }
+        new(user, options)
+      rescue KeyError
```
Don't think this is necessary. 🛡 ⚔️ 

It would suggest we've allowed bad data into the system to begin with(as its session retrieval).

For the `options` we could replace the fetches with a `slice`.

dan-palmer commented on app/models/sso/identity.rb:
```diff
@@ -0,0 +1,62 @@
+        user = User.from_omniauth(auth_hash)
+        options = extract_additional_info(info)
+        new(user, options)
+      rescue
```
This seems to be an exceptional situation

dan-palmer commented on config/initializers/devise.rb:
```diff
@@ -1,317 +0,0 @@
-  # ==> Configuration for :timeoutable
-  # The time you want to timeout the user session without activity. After this
-  # time the user will be asked for credentials again. Default is 30 minutes.
-  config.timeout_in = 60.minutes
```
Devise provided a timeout. Should we add this to the `session_store`?

dan-palmer commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+  helper_method :sso_path
+
+  def build_url(url, redirect_url = nil)
```
Think this shouldn't be generic & should just be `def logout_url` with no args.

dan-palmer commented on app/controllers/application_controller.rb:
```diff
@@ -24,4 +24,30 @@ def redirect_unless_document_editable
+
+  def set_sso_identity
+    session[:sso_data] && SSO::Identity.from_session(session[:sso_data])
+  rescue SSO::Identity::InvalidData
```
Do we have a test for this?

dan-palmer commented on app/assets/stylesheets/moving_people_safely/_authentication.scss:
```diff
@@ -1,9 +1,12 @@
-.sign-out {
+#signed-in-box {
```
OMG 🍼 🍺

Commit:
Minor tweaks to the SSO integration

- Add a session timeout setting, resets once the session expired
- Simplify some of the SSO data handling

lostie commented:
👍

dan-palmer commented:
Beautiful 👸 🤴
